ts/Data/Connectors/JSONConnector.ts

```diff
@@ -0,0 +1,243 @@
+ *
+ * */
+
+'use strict';
+
+/* *
+ *
+ *  Imports
+ *
+ * */
+
+import type DataEvent from '../DataEvent';
+import type Types from '../../Shared/Types';
+import type JSONConnectorOptions from './JSONConnectorOptions';
+
+import DataConnector from './DataConnector.js';
+import U from '../../Core/Utilities.js';
+import JSONConverter from '../Converters/JSONConverter.js';
+const { merge } = U;
+
+/* *
+ *
+ *  Class
+ *
+ * */
+
+/**
+ * Class that handles creating a DataConnector from JSON structure
+ *
+ * @private
+ */
+class JSONConnector extends DataConnector {
+    /* *
+     *
+     *  Static Properties
+     *
+     * */
+
+    protected static readonly defaultOptions: JSONConnectorOptions = {
+        data: [],
+        enablePolling: false,
+        dataRefreshRate: 0,
+        firstRowAsNames: true,
+        orientation: 'rows'
+    };
+
+    /* *
+     *
+     *  Constructor
+     *
+     * */
+
+    /**
+     * Constructs an instance of JSONConnector.
+     *
+     * @param {JSONConnector.UserOptions} [options]
+     * Options for the connector and converter.
+     */
+    public constructor(
+        options?: JSONConnector.UserOptions
+    ) {
+        const mergedOptions = merge(JSONConnector.defaultOptions, options);
+
+        super(mergedOptions);
+
+        this.converter = new JSONConverter(mergedOptions);
+        this.options = mergedOptions;
+
+        if (mergedOptions.enablePolling) {
+            this.startPolling(
+                Math.max(mergedOptions.dataRefreshRate || 0, 1) * 1000
+            );
+        }
+    }
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    /**
+     * Options related to the handling of the JSON DataConnector,
+     * i.e. source, fetching, polling
+     */
+    public readonly options: JSONConnectorOptions;
+
+    /**
+     * The attached parser that converts the data format to the table.
+     */
+    public readonly converter: JSONConverter;
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    /**
+     * Initiates the loading of the JSON source to the connector
+     *
+     * @param {DataEvent.Detail} [eventDetail]
+     * Custom information for pending events.
+     *
+     * @emits JSONConnector#load
+     * @emits JSONConnector#afterLoad
+     */
+    public load(eventDetail?: DataEvent.Detail): Promise<this> {
+        const connector = this,
+            converter = connector.converter,
+            table = connector.table,
+            { data, dataUrl, dataModifier } = connector.options;
+
+        connector.emit<JSONConnector.Event>({
+            type: 'load',
+            data,
+            detail: eventDetail,
+            table
+        });
+
+        // If already loaded, clear the current rows
+        table.deleteRows();
+
+        return Promise.resolve(
+            data ?
+                data :
+                dataUrl ?
+                    fetch(dataUrl).then((json): Promise<any> => json.json()) :
+                    []
+        )
```

Comment:
Can you change the formatting, its really difficult to read, because of intends, and nested conditions.